### ISSUE TITLE
Latency degradation for test_latency_write & test_latency_mixed

### ISSUE THREAD

**Issue by roydahan** [scylladb/scylladb#10981]
Version: 5.0.0
Version Details:
build date: 20220628
commit id: c9e79cb4a
build-id: daef393310f38c086a6e5070c8f2013684973a48
Setup Details:
ami_id_db_scylla: ami-0ae3b0658c86b30cd
cluster_backend: aws
instance_type_db: i3.4xlarge
instance_type_loader: c4.2xlarge
instance_type_monitor: t3.large

Version | Test_write - P99 (ms) | Test_read - P99 (ms) | Test_mixed - P99 (ms)
-- | -- | -- | --
5.0.0 | 6.7 | 3.1 | 7.4
5.0.rc8 | 6.7 | 3.2 | 7.0
5.0.rc5 | 6.0 | 3.2 | 7.6
5.0.rc3 | 7.4 | 3.1 | 6.3
5.0.rc2 | 5.2 | 3.2 | 6.2
5.0.rc1 | 4.2 | 2.8 | 4.2
2022.1.rc8 | 6.1 | 3.4 | 6.6
4.6.3 | 4.1 | 3.5 | 4.5
4.6.rc5 | 4.1 | 3.2 | 5.4
4.5.rc7 | 3.7 | 3.2 | 4.4
4.4.1 | 3.3 | 2.6 | 6.0
4.3.0 | 2.8 | 2.6 | 5.6

**Comment by roydahan** [comment]
I'm starting to bisect with 5.0.rc1 (hoping it's a stable enough build).

**Comment by roydahan** [comment]
@slivne / @eliransin, I used 5.0 RCs to try bisecting the regression (and added the information to the original table).
The results also support my previous comment.
**RC1 is the last good result** we have for WRITE and MIXED.

Although RC2 isn't bad as the next ones, I think we should start looking at it.

**Comment by roydahan** [comment]
f92622e0d - **(tag: scylla-5.0.rc3)** release: prepare for 5.0.rc3 (3 months ago) <Yaron Kaikov>
 3bca608db - docker: run scylla as root (3 months ago) <Takuya ASADA>
 a93b72d5d - docker: revert scylla-server.conf service name change (3 months ago) <Takuya ASADA>
 d58ca2edb - range_tombstone_list: insert_from: correct rev.update range_tombstone in not overlapping case (3 months ago) <Benny Halevy>
 75740ace2 - dist/docker: fix incorrect locale value (3 months ago) <Alexey Kartashov>
 d7a1bf633 - cql3: fix qualifying restrictions with IN for indexing (3 months ago) <Piotr Sarna>
 bbd7d657c - Update seastar submodule (pidof command not installed) (3 months ago) <Avi Kivity>
 f5bf4c81d - Merge 'replica/database: truncate: temporarily disable compaction on table and views before flush' from Benny Halevy (4 months ago) <Avi Kivity>
 02e833665 - atomic_cell: compare_atomic_cell_for_merge: compare ttl if expiry is equal (4 months ago) <Benny Halevy>
 601812e11 - atomic_cell: compare_atomic_cell_for_merge: compare ttl if expiry is equal (4 months ago) <Benny Halevy>
 ea466320d - atomic_cell: compare_atomic_cell_for_merge: fixup indentation (4 months ago) <Benny Halevy>
 25ea831a1 - atomic_cell: compare_atomic_cell_for_merge: simplify expiry/deltion_time comparison (4 months ago) <Benny Halevy>
 8648c79c9 - main: shutdown: do not abort on certain system errors (4 months ago) <Benny Halevy>
 7ae4d0e6f - Seastar: backport Seastar fix for missing scring escape in JSON output (4 months ago) <Nadav Har'El>
 f3564db94 - expression: fix get_value for mismatched column definitions (4 months ago) <Piotr Sarna>
 97caf1283 - Update seastar submodule (IO preemption overlap) (4 months ago) <Pavel Emelyanov>
 839d9ef41 - **(tag: scylla-5.0.rc2)** release: prepare for 5.0.rc2 (4 months ago) <Yaron Kaikov>
 782bd50f9 - compaction_manager: rewrite_sstables: do not acquire table write lock (4 months ago) <Benny Halevy>
 0a4d971b4 - Merge 'utils: cached_file: Fix alloc-dealloc mismatch during eviction' from Tomasz Grabiec (5 months ago) <Avi Kivity>
 22562f767 - cql3: result_set: remove std::ref from comperator& (5 months ago) <Benny Halevy>
 eb80dd1db - Revert "sstables/compaction_manager: rewrite_sstables(): resolve maintenance group FIXME" (5 months ago) <Raphael S. Carvalho>
 51d699ee2 - Update seastar submodule (overzealous log silencer) (5 months ago) <Avi Kivity>
 83a33bff8 - Point seastar submodule at scylla-seastar.git (5 months ago) <Avi Kivity>
 273563b9a - alternator: allow REMOVE of non-existent nested attribute (5 months ago) <Nadav Har'El>
 891990ec0 - **(tag: scylla-5.0.rc1)** release: prepare for 5.0.rc1 (5 months ago) <Yaron Kaikov>

**Comment by slivne** [comment]
@avikivity - we discussed this on the call need to bisect between 5.0.rc1 - 5.0.rc3 and there is a need todo backports.

**Comment by avikivity** [comment]
It's possible the regression is related to the kernel version too.

**Comment by roydahan** [comment]
> It's possible the regression is related to the kernel version too.

Indeed possible, I just checked it and I see that:
rc1 has "5.11.0-1028-aws"
While 
rc2-rc8 have "5.13.0-10(?)(?)-aws"

**Comment by eliransin** [comment]
@avikivity @roydahan 
Just to see that we are aligned on the method of operation here:
1. I will build RC3 using the old kernel. - if running it doesn't fix the regression we will proceed to step 2.
2. I am going to make a new Branch with a modified history (where the Backport form @avikivity is applied in the scylla-5.0.rc1 tag. Then I will make some evenly spaced images all build with the same machine image (rc1) so it will eliminate the kernel as a variant. Roy will test and then we will have two-three more bisection stages for nailing the ofending commit down.

Sounds good?

**Comment by roydahan** [comment]
Sounds good.
I tried to run a test where we start with RC1 (old kernel) and upgrade scylla to RC2, but it gets stuck during the upgrade and I didn't have time to debug it.

**Comment by slivne** [comment]
@michoecho we need help here i am assigning this to you - @roydahan will help you ping him. 

/cc @avikivity

**Comment by michoecho** [comment]
@roydahan Which test is it? `test-cases/performance/perf-regression-latency-1TB.yaml`?

I'm not a SCT user. Should I know something undocumented before I attempt to run it?

**Comment by roydahan** [comment]
> @roydahan Which test is it? `test-cases/performance/perf-regression-latency-1TB.yaml`?
Yes.
> 
> I'm not a SCT user. Should I know something undocumented before I attempt to run it?

All details should be in the code of "test_latency" and the yaml you mentioned above.
If you want to run it with official version and just look at a monitor or something like that, you can do it from this Jenkins job:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/scylla-master-perf-regression-aws-test-latency/

**Comment by michoecho** [comment]
Okay, I have finished my investigation.

I couldn't get SCT to work well, and the jenkins job which originally performed this test is broken due to some dependency problems (https://jenkins.scylladb.com/job/Perf-Regression/job/scylla-release-perf-regression-aws-test-latency/72/), so I wrote my own set of scripts to reproduce the workload (https://github.com/michoecho/scylla-cluster-test-scripts). Using them, not only did I see no regression between 5.0.rc1 and 5.0.rc3, but I saw that both give significantly better results than what is given in the issue for pre-regression versions. After lots of digging, I found out that the meaningful difference between my setup and the SCT setup is `--blocked-reactor-notify-ms 4`.

As it turns out, this latency benchmark has been broken since 71d721a97e20df2c6b0009f7acc6b88dae4b8c9f (introduced in May 2020 in the 4.6 cycle). This commit adds logger calls (quite fancy ones, with backtraces and all) whenever memory reclaim takes more than `blocked-reactor-notify-ms`. However, this mechanism uses `CLOCK_MONOTONIC_COARSE` to measure the duration of reclaim. As it happens, the resolution of `CLOCK_MONOTONIC_COARSE` on the test i3 machines is 4000 µs (:man_facepalming:), equal to the test's `--blocked-reactor-notify-ms`, so a reclaim with any duration is able to trigger the log, if it happens during the coarse tick.

This causes the database to spam gigabytes of `lsa-timing` logs during affected tests, which significantly degrades tail latencies. The impact on p99 is as seen in this issue. The impact on higher quantiles is much more dramatic — if we reported higher quantiles, this regression would be visible immediately.
(To see the log spam, you can e.g. look into the database logs of https://jenkins.scylladb.com/view/master/job/scylla-master/job/scylla-master-perf-regression-aws-test-latency/207)

Nefariously, this phenomenon doesn't always occur, because the clock period varies _just_ slightly between machines (I don't know why). If it's below 4000 µs, then the log spam doesn't happen. E.g. in the test run where I noticed this, only one of the 3 machines was spamming the log; the other two had a clock period of ~3999 µs (:man_facepalming:) and didn't do it. This explains the instability of results in post-4.6 runs. (I can't confirm directly that the "better" runs had less spamming machines since relevant jenkins logs have already disappeared. Not sure if there's a point, though).

To sum up, I claim that the results of `scylla-release-perf-regression-aws-test-throughput` and `scylla-master-perf-regression-aws-test-latency` have been noise for the last 16 months. (Tests with longer `blocked-reactor-notify-ms`, such as the latency under RBNO tests — which IIUC have `blocked-reactor-notify-ms=5` — shouldn't be affected.)

**Comment by roydahan** [comment]
@michoecho thanks for the detailed explanation.
It does make a lot of sense.
I'm sorry that you couldn't run SCT or the jobs directly, next time please ping us so we may help with that.

What about scylla side of fancy logging? can't we fix that?

@avikivity / @slivne shall we run these tests with a higher stall detector?
(It was lowered few years ago to ease degradations investigations, requested by dev).

**Comment by avikivity** [comment]
> Okay, I have finished my investigation.
> 
> I couldn't get SCT to work well, and the jenkins job which originally performed this test is broken due to some dependency problems (https://jenkins.scylladb.com/job/Perf-Regression/job/scylla-release-perf-regression-aws-test-latency/72/), so I wrote my own set of scripts to reproduce the workload (https://github.com/michoecho/scylla-cluster-test-scripts). Using them, not only did I see no regression between 5.0.rc1 and 5.0.rc3, but I saw that both give significantly better results than what is given in the issue for pre-regression versions. After lots of digging, I found out that the meaningful difference between my setup and the SCT setup is `--blocked-reactor-notify-ms 4`.
> 
> As it turns out, this latency benchmark has been broken since [71d721a](https://github.com/scylladb/scylladb/commit/71d721a97e20df2c6b0009f7acc6b88dae4b8c9f) (introduced in May 2020 in the 4.6 cycle). This commit adds logger calls (quite fancy ones, with backtraces and all) whenever memory reclaim takes more than `blocked-reactor-notify-ms`. However, this mechanism uses `CLOCK_MONOTONIC_COARSE` to measure the duration of reclaim. As it happens, the resolution of `CLOCK_MONOTONIC_COARSE` on the test i3 machines is 4000 µs (man_facepalming), equal to the test's `--blocked-reactor-notify-ms`, so a reclaim with any duration is able to trigger the log, if it happens during the coarse tick.

Wow, :man_facepalming: indeed. Good job tracking it down. Changes to the test harness are always much harder to pick out.

Please patch it to require at least two ticks of the clock (one tick may be too low, we can have a reclaim start just before a tick boundary and end just after it).

**Comment by avikivity** [comment]
> @michoecho thanks for the detailed explanation. It does make a lot of sense. I'm sorry that you couldn't run SCT or the jobs directly, next time please ping us so we may help with that.
> 
> What about scylla side of fancy logging? can't we fix that?
> 
> @avikivity / @slivne shall we run these tests with a higher stall detector? (It was lowered few years ago to ease degradations investigations, requested by dev).

It's a bug in ScyllaDB that it reports much small stalls (even a nanosecond will be reported), so we need to fix it in ScyllaDB.

**Comment by michoecho** [comment]
> Please patch it to require at least two ticks of the clock (one tick may be too low, we can have a reclaim start just before a tick boundary and end just after it).

@avikivity I posted a patch to https://github.com/scylladb/scylladb/pull/11680 which does something slightly different: it requires the measured duration to be at least 1 tick bigger than the stall threshold. This also prevents false positives for thresholds bigger than a tick. (E.g. if we just require two ticks, if resolution is 4ms resolution and threshold is 8ms, a 5ms reclaim will be measured at 8ms and will trigger the warning. If we require the duration to be at least threshold + resolution, then there are no false positives). Of course, this also removes some true positives (but only those which are impossible to be determined as true or false).

**Comment by roydahan** [comment]
@avikivity do you want to backport this to any of the enterprise releases?
I know it shouldn't affect anyone who doesn't change the blocked-reactor-notify-ms, but just checking.

**Comment by mykaul** [comment]
> @avikivity do you want to backport this to any of the enterprise releases? I know it shouldn't affect anyone who doesn't change the blocked-reactor-notify-ms, but just checking.

If it's fixed in 5.1, I rather not.

**Comment by roydahan** [comment]
It's not in 5.1, I asked for enterprise releases because they live longer and the results of our latency tests are skewed due to this issue, but we probably shouldn't be worried about it.

**Comment by mykaul** [comment]
@avikivity - ping - do we need this backported to 5.1?

**Comment by avikivity** [comment]
We don't - this can only happen with non-default reactor stall threshold, which is rare outside tests.
